docs/design_docs/011-historacle-pricing.md

```diff
@@ -0,0 +1,88 @@
+# Design Doc 011: Historacle Pricing
+
+## Status
+
+Draft
+
+## Abstract
+
+In order to support small-volume assets in a safe manner, we need to be able to calculate the weighted medians of our prices over the past X amount of time (30 days or so). This will allow the leverage module to make decisions around when to allow liquidations if an asset price is considered abnormal.
+
+## Context
+
+Right now, the volume of assets such as Juno, Umee, and Osmo are so low that it would be relatively easy to use the leverage protocol on UMEE to manipulate the prices of these assets and game the system.
+
+The attack goes:
+
+1. Account A deposits & collateralizes a large amount of USDC.
+2. Account A uses collateral to borrow a lot of the low-volume token (here, FOO).
+3. Account A sells the tokens to Account B, inflating the price of FOO.
+4. Account B collateralizes inflated FOO, and borrows USDC.
+5. Account B exits the market for a profit.
+
+In order to avoid this, and to continue with our goal of allowing users to collateralize low-volume assets, we need to have a safety net.
+
+Currently, this is defended against by disallowing the use of `Umee` as collateral, however, we would like to re-enable that and list other assets with low volume for collateral.
+
+Historacle pricing will provide an API for the leverage module to tell when prices have changed in an abnormal fashion and avoid such events.
+
+## Specification
+
+Each `Stamp Period`, the `x/oracle` module will now "stamp" the set of exchange rates in memory, and store it until a `Pruning Period` has passed (30 days). Another epoch, `Median Period`, will determine how often the Median and the `Standard Deviation around the Median` are calculated, which will also be stored in memory.
+
+These values are stored in state in order to avoid the `x/leverage` module from having to calculate them while making decisions around allowable positions for users to take.
+
+### Proposed API
+
+The leverage protocol will have access to a `Keeper` function from the `x/oracle` module, which will take a `denom`, and will return whether or not the current price of that asset is within the `Standard Deviation around the Median`.
+
```

Comment:
```suggestion
The leverage protocol will have access to the following `keeper` functions from the `x/oracle` module:
- `historicalMedian(denom) (sdk.Dec, error)` returns the median price of an asset in the last `Pruning Period`
- `withinStandardDeviation(denom) (bool, error)` returns whether or not the current price of an asset is within the `Standard Deviation around the Median`.

```
Is this still the plan?

Reply:
Do you need `historicalMedian` @toteki ? How else are you going to use it other than using `withinStandardDeviation`?

Reply:
The current plan involves ensuring any `MsgBorrow`, `MsgWithdraw`, or `MsgDecollateralize` results in a position that is healthy (under borrow limit) using both current an historical median prices.

For that I need the specific price - the other option (if I just have a boolean `withinStandardDeviation`) would be to disable all three message types, even for tiny borrow/etc amounts, if one of a user's borrowed or collateral assets  returned `false`.

Reply:
Gotcha, we'll provide both APIs for ya then 🚀

Reply:
Please see 04c4082c48d642804c438c18de8d985b6850cfec, changed the function names a bit